fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMergeTest.java

```diff
@@ -197,6 +202,185 @@ void testMergeTimeout() {
+        final String indexType = "nonRetriableExceptionIndex";
+        AtomicInteger attemptCount = new AtomicInteger(0);
+
+        TestFactory.register(indexType, state -> {
+            adjustMergeControl(state);
+            attemptCount.incrementAndGet();
+
+            // Throw a RuntimeException that is not retriable
+            final CompletableFuture<Void> future = new CompletableFuture<>();
+            future.completeExceptionally(new IllegalStateException("Non-retriable error"));
+            return future;
+        });
+
+        final FDBRecordStore.Builder storeBuilder = createStore(indexType);
+        try (OnlineIndexer indexer = OnlineIndexer.newBuilder()
+                .setRecordStoreBuilder(storeBuilder)
+                .setTargetIndexesByName(List.of(INDEX_NAME))
+                .setMaxAttempts(10)
+                .build()) {
+            Assertions.assertThrows(IllegalStateException.class, indexer::mergeIndex);
+        }
+
+        // Should only attempt once, no retries
+        assertEquals(1, attemptCount.get());
+    }
+
+    /**
+     * Test that a TimeoutException causes retry behavior (not abort).
+     */
+    @ParameterizedTest
+    @BooleanSource
+    void testTimeoutExceptionRetries(boolean customTimeout) {
+        final String indexType = "timeoutExceptionIndex";
+        final String exceptionMessage = "my timeout";
+        AtomicInteger attemptCount = new AtomicInteger(0);
+
+        TestFactory.register(indexType, state -> {
+            adjustMergeControl(state);
+
+            attemptCount.incrementAndGet();
+
+            // Throw a TimeoutException (not FDB timeout)
+            final CompletableFuture<Void> future = new CompletableFuture<>();
+            future.completeExceptionally(
+                    customTimeout ?
+                    new CustomOperationTimeoutException(exceptionMessage) :
+                    new java.util.concurrent.TimeoutException(exceptionMessage));
+            return future;
+        });
+
+        final FDBRecordStore.Builder storeBuilder = createStore(indexType);
+        try (OnlineIndexer indexer = OnlineIndexer.newBuilder()
+                .setRecordStoreBuilder(storeBuilder)
+                .setTargetIndexesByName(List.of(INDEX_NAME))
+                .setMaxAttempts(5)
+                .build()) {
+            Exception thrownException = Assertions.assertThrows(Exception.class, indexer::mergeIndex);
+            // Assert that the timeout exception is in the cause chain
+            var timeoutCause =
+                    customTimeout ?
+                    findCause(thrownException, CustomOperationTimeoutException.class) :
+                    findCause(thrownException, TimeoutException.class);
+            Assertions.assertNotNull(timeoutCause);
+            Assertions.assertEquals(exceptionMessage, timeoutCause.getMessage());
+        }
+        // Assert multiple retries
+        assertTrue(1 < attemptCount.get());
```

Comment:
Do we know exactly how many retries should be? Isn't it the same as the number above (5)?